### ISSUE TITLE
"network is unreachable" attempting to pull from registry over 6PLANE from WSL2

### ISSUE THREAD

**Issue by M-Pixel** [moby/moby#41981]
**Description**
I am unable to pull from an IPv6-only registry on Windows, but I am able to pull from it on Ubuntu.

**Steps to reproduce the issue:**
1. Start with three computers, two running Ubuntu Bionic, one running Windows 10 20H2, all running the latest Docker.
2. Network the three computers together so that they can all ping each other.
3. Configure the network for 6PLANE.  Connecting them over a ZeroTier-One VPN is probably the easiest way to accomplish this.
4. On Ubuntu A, compose a registry like this

```
version: '3'
services:
  registry:
    build: ./
    restart: unless-stopped
    networks:
      zerotier:
        ipv6_address: 'F00:BA::1'
networks:
  zerotier:
    enable_ipv6: true
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 'F00:BA::/96'
```

configured like this

```
version: 0.1
storage:
  inmemory: {}
http:
  addr: '[F00:BA::1]:5000'
```

... substituting, of course, a valid IP address and subnet for your network and host.

5. Double-check that you can ping F00::BA::1 from the other computers
6. On Ubuntu B and Windows, add "F00:BA::1 docker-registry" to the hosts file
7. On Ubuntu B, execute `docker pull docker-registry:5000/foo`

**Describe the results you received:**
Ubuntu: `Error response from daemon: manifest for docker-registry:5000/foo:latest not found: manifest unknown: manifest unknown`

Windows: `Error response from daemon: Get http://docker-registry:5000/v2/: dial tcp [f00:ba::1]:5000: connect: network is unreachable`

**Describe the results you expected:**
Windows returns the exact same message that Ubuntu did ("manifest unknown").

**Output of `docker version`:**

Ubuntu
```
Client:
 Version:           19.03.6
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        369ce74a3c
 Built:             Fri Dec 18 12:21:44 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.6
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       369ce74a3c
  Built:            Thu Dec 10 13:23:49 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.3-0ubuntu1~18.04.4
  GitCommit:
 runc:
  Version:          spec: 1.0.1-dev
  GitCommit:
 docker-init:
  Version:          0.18.0
  GitCommit:
```

Windows"
```
Client: Docker Engine - Community
 Cloud integration: 1.0.7
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:14:16 2020
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:15:28 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

Ubuntu
```
Client:
 Debug Mode: false

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 4
 Server Version: 19.03.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version:
 runc version:
 init version:
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.15.0-135-generic
 Operating System: Ubuntu 18.04.5 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.615GiB
 Name: openio1
 ID: OINM:QNCY:VGBL:2HZV:HHOM:KS5L:PTSN:Q64O:U6US:ZMAV:PIZG:YE6S
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

Windows
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.5.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 19
 Server Version: 20.10.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.128-microsoft-standard
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 50.16GiB
 Name: docker-desktop
 ID: WX2C:IDMF:THHV:4OSQ:SN3I:MA2N:FIN6:R3YU:EBDR:UC5B:TQ2H:PSQH
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 38
  Goroutines: 41
  System Time: 2021-02-04T05:19:17.528462Z
  EventsListeners: 2
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio weight support
WARNING: No blkio weight_device support
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Comment by thaJeztah** [comment]
@stephen-turner ptal

**Comment by stephen-turner** [comment]
Unfortunately IPv6 is not fully working in Docker Desktop. https://docs.docker.com/docker-for-windows/troubleshoot/#networking-issues. There is a roadmap item at https://github.com/docker/roadmap/issues/156 which you can add a thumbs-up to if you want us to prioritise it.

**Comment by thaJeztah** [comment]
Thanks @stephen-turner 

> There is a roadmap item at docker/roadmap#156 which you can add a thumbs-up to if you want us to prioritise it.

I think that issue is slightly different; the docker engine itself already has (partial) support for ipv6; in the WSL2 situation, the engine running on Windows is a Linux binary (so "by itself") should provide the same functionality as the "Ubuntu" case.

I guess there's differences in the "network stack" when using WSL2 as on a "native" Linux host; perhaps a separate roadmap issue would be needed (or more details added to the existing roadmap issue).

That said; @M-Pixel I now notice that the Ubuntu example uses an older version of the docker engine than the WSL2 version; I know there have been changes in IPv6 in Docker 20.10 (and there were some regressions that will be fixed in an upcoming patch release); I don't think those changes should affect _pulling images_, but to be sure; if you have a Ubuntu machine to test on, could you try if things work "as expected" on Ubuntu with Docker 20.10.2?

**Comment by M-Pixel** [comment]
I was able to test this out using Docker 20.10.3 on Ubuntu 20.10 (live CD inside a Hyper-V machine on the very same Windows machine).  It exhibited the correct behavior, so the problem definitely correlates to Windows/WSL2 rather than to Docker 20.  I suppose it's possible that it could be a .2 bug that was fixed in .3, as I couldn't find a way to install 20.10.2 on Ubuntu, nor a way of updating to 20.10.3 on Windows, but I'd be surprised.
